File: src/main/java/org/cubexmc/metro/gui/view/LineListView.java
package org.cubexmc.metro.gui.view;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.gui.GuiColors;
import org.cubexmc.metro.gui.GuiHolder;
import org.cubexmc.metro.gui.GuiSlots;
import org.cubexmc.metro.gui.ItemBuilder;
import org.cubexmc.metro.manager.LanguageManager;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.model.Stop;
import org.cubexmc.metro.util.OwnershipUtil;

public final class LineListView {
    private final Metro plugin;

    public LineListView(Metro plugin) {
        this.plugin = plugin;
    }

    public void openLineList(Player player, int requestedPage, boolean showOnlyMine,
                             GuiHolder.GuiView previousView) {
        GuiHolder holder = createHolder(GuiHolder.GuiType.LINE_LIST, previousView);
        holder.setData("page", requestedPage);
        holder.setData("showOnlyMine", showOnlyMine);

        List<Line> allLines = new ArrayList<>(plugin.getLineManager().getAllLines());
        List<Line> filteredLines;
        if (showOnlyMine && !OwnershipUtil.hasAdminBypass(player)) {
            filteredLines = allLines.stream()
                    .filter(line -> OwnershipUtil.canManageLine(player, line))
                    .collect(Collectors.toList());
        } else {
            filteredLines = allLines;
        }

        Map<String, List<Line>> groupedLines = filteredLines.stream()
                .collect(Collectors.groupingBy(Line::getName));
        List<String> sortedNames = new ArrayList<>(groupedLines.keySet());
        sortedNames.sort(String::compareTo);

        holder.setData("lineNames", sortedNames);
        holder.setData("groupedLines", groupedLines);

        String titleKey = showOnlyMine ? "gui.line_list.title_mine" : "gui.line_list.title_all";
        String title = ChatColor.translateAlternateColorCodes('&', msg(titleKey));
        renderPaginatedList(player, holder, title, sortedNames, requestedPage, (inv, name, slot) -> {
            List<Line> variants = groupedLines.get(name);
            Line representative = variants.get(0);

            List<String> lore = new ArrayList<>();
            if (variants.size() > 1) {
                lore.add(msg("gui.common.variants", "count", String.valueOf(variants.size())));
                lore.add("");
                lore.add(msg("gui.line_list.click_view_variants"));
            } else {
                lore.add(msg("gui.common.id", "id", representative.getId()));
                lore.add(msg("gui.line_list.stop_count", "count",
                        String.valueOf(representative.getOrderedStopIds().size())));
                addLineSummaryLore(lore, representative);
                if (representative.getColor() != null) {
                    lore.add(msg("gui.line_list.color") + representative.getColor() + "■■■■■");
                }
                lore.add("");
                if (OwnershipUtil.canManageLine(player, representative)) {
                    lore.add(msg("gui.line_list.can_manage"));
                } else {
                    lore.add(msg("gui.line_list.view_only"));
                }
                lore.add("");
                lore.add(msg("gui.line_list.click_view"));
                if (OwnershipUtil.canManageLine(player, representative)) {
                    lore.add(msg("gui.line_list.click_settings"));
                }
            }

            inv.setItem(slot, new ItemBuilder(GuiColors.getWoolByColor(representative.getColor()))
                    .name((representative.getColor() != null ? representative.getColor() : "&f") + name)
                    .lore(lore)
                    .build());
        }, (inv, page, totalPages) -> addControlBar(inv, page, totalPages, showOnlyMine));
    }

    public void openLineVariants(Player player, String lineName, int requestedPage,
                                 GuiHolder.GuiView previousView) {
        GuiHolder holder = createHolder(GuiHolder.GuiType.LINE_VARIANTS, previousView);
        holder.setData("page", requestedPage);
        holder.setData("lineName", lineName);

        List<Line> variants = plugin.getLineManager().getAllLines().stream()
                .filter(line -> line.getName().equals(lineName))
                .sorted(Comparator.comparing(Line::getId))
                .collect(Collectors.toList());
        holder.setData("lines", variants);

        String title = ChatColor.translateAlternateColorCodes('&',
                lineName + " - " + msg("gui.common.variants_title"));
        renderPaginatedList(player, holder, title, variants, requestedPage, (inv, line, slot) -> {
            List<String> lore = new ArrayList<>();
            lore.add(msg("gui.common.id", "id", line.getId()));
            lore.add(msg("gui.line_list.stop_count", "count", String.valueOf(line.getOrderedStopIds().size())));
            addLineSummaryLore(lore, line);
            if (line.getColor() != null) {
                lore.add(msg("gui.line_list.color") + line.getColor() + "■■■■■");
            }
            lore.add("");
            if (OwnershipUtil.canManageLine(player, line)) {
                lore.add(msg("gui.line_list.can_manage"));
            } else {
                lore.add(msg("gui.line_list.view_only"));
            }
            lore.add("");
            lore.add(msg("gui.line_list.click_view"));

            inv.setItem(slot, new ItemBuilder(GuiColors.getWoolByColor(line.getColor()))
                    .name((line.getColor() != null ? line.getColor() : "&f") + line.getId())
                    .lore(lore)
                    .build());
        }, this::addVariantNavigationControls);
    }

    private GuiHolder createHolder(GuiHolder.GuiType type, GuiHolder.GuiView previousView) {
        GuiHolder holder = new GuiHolder(type);
        holder.setPreviousView(previousView);
        return holder;
    }

    private <T> void renderPaginatedList(Player player, GuiHolder holder, String title, List<T> items,
                                         int requestedPage, ItemPopulator<T> populator,
                                         ControlBarPopulator controlBar) {
        int totalPages = Math.max(1, (int) Math.ceil((double) items.size() / GuiSlots.ITEMS_PER_PAGE));
        int page = Math.max(0, Math.min(requestedPage, totalPages - 1));
        holder.setData("page", page);
        holder.setData("totalPages", totalPages);

        Inventory inv = Bukkit.createInventory(holder, 54, title);
        holder.setInventory(inv);

        int start = page * GuiSlots.ITEMS_PER_PAGE;
        int end = Math.min(start + GuiSlots.ITEMS_PER_PAGE, items.size());
        for (int i = start; i < end; i++) {
            populator.populate(inv, items.get(i), i - start);
        }

        controlBar.populate(inv, page, totalPages);
        player.openInventory(inv);
    }

    private void addControlBar(Inventory inv, int page, int totalPages, boolean showOnlyMine) {
        ItemStack filler = new ItemBuilder(Material.GRAY_STAINED_GLASS_PANE)
                .name(" ")
                .build();
        for (int i = GuiSlots.ITEMS_PER_PAGE; i < 54; i++) {
            inv.setItem(i, filler);
        }

        if (page > 0) {
            inv.setItem(GuiSlots.SLOT_PREV_PAGE, new ItemBuilder(Material.ARROW)
                    .name(msg("gui.control.prev_page"))
                    .build());
        }
        inv.setItem(GuiSlots.SLOT_BACK, new ItemBuilder(Material.DARK_OAK_DOOR)
                .name(msg("gui.control.back_main"))
                .build());
        inv.setItem(GuiSlots.SLOT_PAGE_INFO, new ItemBuilder(Material.PAPER)
                .name(msg("gui.control.page_info", "current", String.valueOf(page + 1),
                        "total", String.valueOf(totalPages)))
                .build());

        String filterName = showOnlyMine ? msg("gui.control.filter_mine") : msg("gui.control.filter_all");
        String filterLore = showOnlyMine ? msg("gui.control.filter_lore_mine") : msg("gui.control.filter_lore_all");
        inv.setItem(GuiSlots.SLOT_FILTER, new ItemBuilder(Material.HOPPER)
                .name(filterName)
                .lore(filterLore)
                .build());

        if (page < totalPages - 1) {
            inv.setItem(GuiSlots.SLOT_NEXT_PAGE, new ItemBuilder(Material.ARROW)
                    .name(msg("gui.control.next_page"))
                    .build());
        }
    }

    private void addVariantNavigationControls(Inventory inv, int page, int totalPages) {
        inv.setItem(GuiSlots.SLOT_BACK, new ItemBuilder(Material.ARROW)
                .name(msg("gui.common.back"))
                .build());
        if (totalPages <= 1) {
            return;
        }
        if (page > 0) {
            inv.setItem(GuiSlots.SLOT_PREV_PAGE, new ItemBuilder(Material.ARROW)
                    .name(msg("gui.common.prev_page"))
                    .build());
        }
        inv.setItem(GuiSlots.SLOT_PAGE_INFO, new ItemBuilder(Material.PAPER)
                .name(msg("gui.common.page_info", "page", String.valueOf(page + 1),
                        "total", String.valueOf(totalPages)))
                .build());
        if (page < totalPages - 1) {
            inv.setItem(GuiSlots.SLOT_NEXT_PAGE, new ItemBuilder(Material.ARROW)
                    .name(msg("gui.common.next_page"))
                    .build());
        }
    }

    private void addLineSummaryLore(List<String> lore, Line line) {
        lore.add(msg("gui.line_boarding.next_stop", "stop_name", getInitialNextStopName(line)));
        lore.add(msg("gui.line_boarding.terminus", "terminus_name", getTerminusDisplayName(line)));
        lore.add(msg("gui.line_boarding.price", "price", formatTicketPrice(line)));
    }

    private String getInitialNextStopName(Line line) {
        List<String> stopIds = line.getOrderedStopIds();
        if (stopIds.size() < 2) {
            return msg("gui.line_boarding.unknown_stop");
        }
        String nextStopId = line.getNextStopId(stopIds.get(0));
        Stop nextStop = nextStopId == null ? null : plugin.getStopManager().getStop(nextStopId);
        return nextStop == null ? msg("gui.line_boarding.unknown_stop") : nextStop.getName();
    }

    private String getTerminusDisplayName(Line line) {
        if (line.getTerminusName() == null || line.getTerminusName().isBlank()) {
            return msg("line.info_default");
        }
        return line.getTerminusName();
    }

    private String formatTicketPrice(Line line) {
        double price = line.getTicketPrice();
        if (price <= 0) {
            return msg("gui.line_boarding.free");
        }
        return plugin.getTicketService().format(price);
    }

    private String msg(String key) {
        return plugin.getLanguageManager().getMessage(key);
    }

    private String msg(String key, String... replacements) {
        Map<String, Object> args = LanguageManager.args();
        for (int i = 0; i < replacements.length - 1; i += 2) {
            LanguageManager.put(args, replacements[i], replacements[i + 1]);
        }
        return plugin.getLanguageManager().getMessage(key, args);
    }

    @FunctionalInterface
    private interface ItemPopulator<T> {
        void populate(Inventory inv, T item, int slot);
    }

    @FunctionalInterface
    private interface ControlBarPopulator {
        void populate(Inventory inv, int page, int totalPages);
    }
}

File: src/main/java/org/cubexmc/metro/gui/view/LineListView.kt
package org.cubexmc.metro.gui.view

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.cubexmc.metro.Metro
import org.cubexmc.metro.gui.GuiColors
import org.cubexmc.metro.gui.GuiHolder
import org.cubexmc.metro.gui.GuiSlots
import org.cubexmc.metro.gui.ItemBuilder
import org.cubexmc.metro.manager.LanguageManager
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.util.OwnershipUtil
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min

class LineListView(private val plugin: Metro) {

    fun openLineList(player: Player, requestedPage: Int, showOnlyMine: Boolean, previousView: GuiHolder.GuiView?) {
        val holder = createHolder(GuiHolder.GuiType.LINE_LIST, previousView)
        holder.setData("page", requestedPage)
        holder.setData("showOnlyMine", showOnlyMine)

        val allLines = plugin.lineManager.getAllLines()
        val filteredLines =
            if (showOnlyMine && !OwnershipUtil.hasAdminBypass(player)) {
                allLines.filter { OwnershipUtil.canManageLine(player, it) }
            } else {
                allLines
            }

        val groupedLines = filteredLines.groupBy { it.name }
        val sortedNames = groupedLines.keys.sorted()

        holder.setData("lineNames", sortedNames)
        holder.setData("groupedLines", groupedLines)

        val titleKey = if (showOnlyMine) "gui.line_list.title_mine" else "gui.line_list.title_all"
        val title = ChatColor.translateAlternateColorCodes('&', msg(titleKey))
        renderPaginatedList(
            player,
            holder,
            title,
            sortedNames,
            requestedPage,
            { inv, name, slot -> addLineNameItem(player, inv, name, slot, groupedLines) },
            { inv, page, totalPages -> addControlBar(inv, page, totalPages, showOnlyMine) },
        )
    }

    private fun addLineNameItem(
        player: Player,
        inv: Inventory,
        name: String,
        slot: Int,
        groupedLines: Map<String, List<Line>>,
    ) {
        val variants = groupedLines[name] ?: return
        val representative = variants[0]

        val lore = ArrayList<String>()
        if (variants.size > 1) {
            lore.add(msg("gui.common.variants", "count", variants.size.toString()))
            lore.add("")
            lore.add(msg("gui.line_list.click_view_variants"))
        } else {
            lore.add(msg("gui.common.id", "id", representative.id))
            lore.add(msg("gui.line_list.stop_count", "count", representative.orderedStopIds.size.toString()))
            addLineSummaryLore(lore, representative)
            lore.add(msg("gui.line_list.color") + representative.color + COLOR_SWATCH)
            lore.add("")
            if (OwnershipUtil.canManageLine(player, representative)) {
                lore.add(msg("gui.line_list.can_manage"))
            } else {
                lore.add(msg("gui.line_list.view_only"))
            }
            lore.add("")
            lore.add(msg("gui.line_list.click_view"))
            if (OwnershipUtil.canManageLine(player, representative)) {
                lore.add(msg("gui.line_list.click_settings"))
            }
        }

        inv.setItem(
            slot,
            ItemBuilder(GuiColors.getWoolByColor(representative.color))
                .name(representative.color + name)
                .lore(lore)
                .build(),
        )
    }

    fun openLineVariants(player: Player, lineName: String?, requestedPage: Int, previousView: GuiHolder.GuiView?) {
        val holder = createHolder(GuiHolder.GuiType.LINE_VARIANTS, previousView)
        holder.setData("page", requestedPage)
        holder.setData("lineName", lineName)

        val variants =
            plugin.lineManager.getAllLines()
                .filter { it.name == lineName }
                .sortedBy { it.id }
        holder.setData("lines", variants)

        val title =
            ChatColor.translateAlternateColorCodes('&', lineName + " - " + msg("gui.common.variants_title"))
        renderPaginatedList(
            player,
            holder,
            title,
            variants,
            requestedPage,
            { inv, line, slot -> addLineVariantItem(player, inv, line, slot) },
            { inv, page, totalPages -> addVariantNavigationControls(inv, page, totalPages) },
        )
    }

    private fun addLineVariantItem(player: Player, inv: Inventory, line: Line, slot: Int) {
        val lore = ArrayList<String>()
        lore.add(msg("gui.common.id", "id", line.id))
        lore.add(msg("gui.line_list.stop_count", "count", line.orderedStopIds.size.toString()))
        addLineSummaryLore(lore, line)
        lore.add(msg("gui.line_list.color") + line.color + COLOR_SWATCH)
        lore.add("")
        if (OwnershipUtil.canManageLine(player, line)) {
            lore.add(msg("gui.line_list.can_manage"))
        } else {
            lore.add(msg("gui.line_list.view_only"))
        }
        lore.add("")
        lore.add(msg("gui.line_list.click_view"))

        inv.setItem(
            slot,
            ItemBuilder(GuiColors.getWoolByColor(line.color))
                .name(line.color + line.id)
                .lore(lore)
                .build(),
        )
    }

    private fun createHolder(type: GuiHolder.GuiType, previousView: GuiHolder.GuiView?): GuiHolder {
        val holder = GuiHolder(type)
        holder.setPreviousView(previousView)
        return holder
    }

    private fun <T> renderPaginatedList(
        player: Player,
        holder: GuiHolder,
        title: String,
        items: List<T>,
        requestedPage: Int,
        populator: ItemPopulator<T>,
        controlBar: ControlBarPopulator,
    ) {
        val totalPages = max(1, ceil(items.size.toDouble() / GuiSlots.ITEMS_PER_PAGE).toInt())
        val page = max(0, min(requestedPage, totalPages - 1))
        holder.setData("page", page)
        holder.setData("totalPages", totalPages)

        val inv = Bukkit.createInventory(holder, INVENTORY_SIZE, title)
        holder.setInventory(inv)

        val start = page * GuiSlots.ITEMS_PER_PAGE
        val end = min(start + GuiSlots.ITEMS_PER_PAGE, items.size)
        for (index in start until end) {
            populator.populate(inv, items[index], index - start)
        }

        controlBar.populate(inv, page, totalPages)
        player.openInventory(inv)
    }

    private fun addControlBar(inv: Inventory, page: Int, totalPages: Int, showOnlyMine: Boolean) {
        val filler = ItemBuilder(Material.GRAY_STAINED_GLASS_PANE).name(" ").build()
        for (slot in GuiSlots.ITEMS_PER_PAGE until INVENTORY_SIZE) {
            inv.setItem(slot, filler)
        }

        if (page > 0) {
            inv.setItem(
                GuiSlots.SLOT_PREV_PAGE,
                ItemBuilder(Material.ARROW).name(msg("gui.control.prev_page")).build(),
            )
        }
        inv.setItem(
            GuiSlots.SLOT_BACK,
            ItemBuilder(Material.DARK_OAK_DOOR).name(msg("gui.control.back_main")).build(),
        )
        inv.setItem(
            GuiSlots.SLOT_PAGE_INFO,
            ItemBuilder(Material.PAPER)
                .name(msg("gui.control.page_info", "current", (page + 1).toString(), "total", totalPages.toString()))
                .build(),
        )

        val filterName = if (showOnlyMine) msg("gui.control.filter_mine") else msg("gui.control.filter_all")
        val filterLore = if (showOnlyMine) msg("gui.control.filter_lore_mine") else msg("gui.control.filter_lore_all")
        inv.setItem(
            GuiSlots.SLOT_FILTER,
            ItemBuilder(Material.HOPPER).name(filterName).lore(filterLore).build(),
        )

        if (page < totalPages - 1) {
            inv.setItem(
                GuiSlots.SLOT_NEXT_PAGE,
                ItemBuilder(Material.ARROW).name(msg("gui.control.next_page")).build(),
            )
        }
    }

    private fun addVariantNavigationControls(inv: Inventory, page: Int, totalPages: Int) {
        inv.setItem(GuiSlots.SLOT_BACK, ItemBuilder(Material.ARROW).name(msg("gui.common.back")).build())
        if (totalPages <= 1) {
            return
        }
        if (page > 0) {
            inv.setItem(
                GuiSlots.SLOT_PREV_PAGE,
                ItemBuilder(Material.ARROW).name(msg("gui.common.prev_page")).build(),
            )
        }
        inv.setItem(
            GuiSlots.SLOT_PAGE_INFO,
            ItemBuilder(Material.PAPER)
                .name(msg("gui.common.page_info", "page", (page + 1).toString(), "total", totalPages.toString()))
                .build(),
        )
        if (page < totalPages - 1) {
            inv.setItem(
                GuiSlots.SLOT_NEXT_PAGE,
                ItemBuilder(Material.ARROW).name(msg("gui.common.next_page")).build(),
            )
        }
    }

    private fun addLineSummaryLore(lore: MutableList<String>, line: Line) {
        lore.add(msg("gui.line_boarding.next_stop", "stop_name", getInitialNextStopName(line)))
        lore.add(msg("gui.line_boarding.terminus", "terminus_name", getTerminusDisplayName(line)))
        lore.add(msg("gui.line_boarding.price", "price", formatTicketPrice(line)))
    }

    private fun getInitialNextStopName(line: Line): String {
        val stopIds = line.orderedStopIds
        if (stopIds.size < 2) {
            return msg("gui.line_boarding.unknown_stop")
        }
        val nextStopId = line.getNextStopId(stopIds[0]) ?: return msg("gui.line_boarding.unknown_stop")
        val nextStop = plugin.stopManager.getStop(nextStopId)
        return nextStop?.name ?: msg("gui.line_boarding.unknown_stop")
    }

    private fun getTerminusDisplayName(line: Line): String {
        if (line.terminusName.isBlank()) {
            return msg("line.info_default")
        }
        return line.terminusName
    }

    private fun formatTicketPrice(line: Line): String {
        val price = line.ticketPrice
        if (price <= 0) {
            return msg("gui.line_boarding.free")
        }
        return plugin.ticketService.format(price) ?: ""
    }

    private fun msg(key: String): String = plugin.languageManager.getMessage(key)

    private fun msg(key: String, vararg replacements: String): String {
        val args = LanguageManager.args()
        var index = 0
        while (index < replacements.size - 1) {
            LanguageManager.put(args, replacements[index], replacements[index + 1])
            index += 2
        }
        return plugin.languageManager.getMessage(key, args)
    }

    private fun interface ItemPopulator<T> {
        fun populate(inv: Inventory, item: T, slot: Int)
    }

    private fun interface ControlBarPopulator {
        fun populate(inv: Inventory, page: Int, totalPages: Int)
    }

    private companion object {
        const val INVENTORY_SIZE = 54
        const val COLOR_SWATCH = "■■■■■"
    }
}
